File: langs/python/src/guidearch/models/topsis/critical_constraints.py
"""Critical constraints analysis — topsis.md §6."""

from __future__ import annotations

import itertools
from typing import TYPE_CHECKING

from guidearch.models.constraint import (
    ConflictConstraint,
    DependencyConstraint,
    ThresholdConstraint,
)
from guidearch.models.critical_constraint import CriticalConstraintM
from guidearch.models.triangular_fuzzy import TriangularFuzzyM

if TYPE_CHECKING:
    from guidearch.models.scenario import ScenarioM


def _apply_single_constraint(
    candidate: tuple[str, ...],
    constraint: ConflictConstraint | DependencyConstraint | ThresholdConstraint,
    coeff: dict[tuple[str, str], TriangularFuzzyM],
    prop_kind: dict[str, str],
) -> bool:
    """Return True if candidate passes the given single constraint."""
    c_set = set(candidate)

    if isinstance(constraint, DependencyConstraint):
        src_in = constraint.source_alternative_id in c_set
        tgt_in = constraint.target_alternative_id in c_set
        return src_in == tgt_in  # biconditional

    elif isinstance(constraint, ConflictConstraint):
        return not (constraint.alternative_a_id in c_set and constraint.alternative_b_id in c_set)

    else:  # ThresholdConstraint
        contrib = TriangularFuzzyM.zero()
        for a_id in candidate:
            contrib = contrib + coeff[(a_id, constraint.property_id)]
        defuzzed = contrib.lower  # §4.2: lower vertex only
        kind = prop_kind[constraint.property_id]
        if kind == "min":
            if constraint.max is not None and defuzzed > constraint.max:
                return False
        else:
            if constraint.min is not None and defuzzed < constraint.min:
                return False
        return True


def critical_constraints(scenario: ScenarioM) -> tuple[CriticalConstraintM, ...]:
    """Rank constraints by how many candidates they eliminate — topsis.md §6.

    For each constraint k:
    1. Build unconstrained Cartesian product.
    2. Apply only k and count eliminated candidates.

    Returns constraints sorted by eliminated descending (most-binding first).
    """
    if not scenario.constraints:
        return ()

    # Build unconstrained Cartesian product
    alts_by_dec: dict[str, list[str]] = {}
    for a in scenario.alternatives:
        alts_by_dec.setdefault(a.decision_id, []).append(a.id)

    pools: list[list[str]] = [alts_by_dec.get(d.id, []) for d in scenario.decisions]
    unconstrained: list[tuple[str, ...]] = [
        tuple(combo) for combo in itertools.product(*pools)
    ]
    total = len(unconstrained)

    coeff: dict[tuple[str, str], TriangularFuzzyM] = {
        (c.alternative_id, c.property_id): c.value for c in scenario.coefficients
    }
    prop_kind: dict[str, str] = {p.id: p.kind for p in scenario.properties}

    results: list[CriticalConstraintM] = []
    for idx, constraint in enumerate(scenario.constraints):
        passed = sum(
            1
            for c in unconstrained
            if _apply_single_constraint(c, constraint, coeff, prop_kind)
        )
        eliminated = total - passed
        results.append(
            CriticalConstraintM(
                constraint_index=idx,
                kind=constraint.kind,
                eliminated=eliminated,
                total=total,
                redundant=(eliminated == 0),
            )
        )

    # Sort by eliminated descending (most-binding first)
    results.sort(key=lambda r: -r.eliminated)

    return tuple(results)

File: langs/typescript/src/models/topsis/critical-constraints.ts
/**
 * Critical constraints analysis — topsis.md §6.
 */
import type { ScenarioM } from '../scenario.js';
import type { ConstraintM } from '../constraint.js';
import type { CriticalConstraintM } from '../critical-constraint.js';
import { TriangularFuzzyM } from '../triangular-fuzzy.js';

function applySingleConstraint(
  candidate: string[],
  constraint: ConstraintM,
  coeff: Map<string, TriangularFuzzyM>,
  propKind: Map<string, string>,
): boolean {
  const cSet = new Set(candidate);

  if (constraint.kind === 'dependency') {
    const srcIn = cSet.has(constraint.sourceAlternativeId);
    const tgtIn = cSet.has(constraint.targetAlternativeId);
    return srcIn === tgtIn; // biconditional
  } else if (constraint.kind === 'conflict') {
    return !(cSet.has(constraint.alternativeAId) && cSet.has(constraint.alternativeBId));
  } else {
    // threshold
    let contrib = TriangularFuzzyM.zero();
    for (const aId of candidate) {
      contrib = contrib.add(coeff.get(`${aId}|${constraint.propertyId}`)!);
    }
    const defuzzed = contrib.lower; // §4.2: lower vertex only
    const kind = propKind.get(constraint.propertyId)!;
    if (kind === 'min') {
      if (constraint.max !== undefined && defuzzed > constraint.max) return false;
    } else {
      if (constraint.min !== undefined && defuzzed < constraint.min) return false;
    }
    return true;
  }
}

/**
 * Rank constraints by how many candidates they eliminate — topsis.md §6.
 *
 * For each constraint k:
 * 1. Build unconstrained Cartesian product.
 * 2. Apply only k and count eliminated candidates.
 *
 * Returns constraints sorted by eliminated descending (most-binding first).
 */
export function criticalConstraints(scenario: ScenarioM): CriticalConstraintM[] {
  if (!scenario.constraints.length) return [];

  // Build unconstrained Cartesian product
  const altsByDec = new Map<string, string[]>();
  for (const d of scenario.decisions) {
    altsByDec.set(d.id, []);
  }
  for (const a of scenario.alternatives) {
    altsByDec.get(a.decisionId)!.push(a.id);
  }

  const pools = scenario.decisions.map((d) => altsByDec.get(d.id) ?? []);

  function cartesian(pools: string[][]): string[][] {
    if (!pools.length) return [[]];
    const [first, ...rest] = pools;
    const restProduct = cartesian(rest);
    const result: string[][] = [];
    for (const a of first) {
      for (const r of restProduct) {
        result.push([a, ...r]);
      }
    }
    return result;
  }

  const unconstrained = cartesian(pools);
  const total = unconstrained.length;

  const coeff = new Map<string, TriangularFuzzyM>();
  for (const c of scenario.coefficients) {
    coeff.set(`${c.alternativeId}|${c.propertyId}`, c.value);
  }
  const propKind = new Map<string, string>(scenario.properties.map((p) => [p.id, p.kind]));

  const results: CriticalConstraintM[] = scenario.constraints.map((constraint, idx) => {
    const passed = unconstrained.filter((c) =>
      applySingleConstraint(c, constraint, coeff, propKind),
    ).length;
    const eliminated = total - passed;
    return {
      constraintIndex: idx,
      kind: constraint.kind,
      eliminated,
      total,
      redundant: eliminated === 0,
    };
  });

  // Sort by eliminated descending (most-binding first)
  results.sort((a, b) => b.eliminated - a.eliminated);

  return results;
}
